src/bms/ast/ast_parse.rs

```diff
@@ -18,12 +20,22 @@ pub(super) fn parse_control_flow_ast<'a>(
             Unit::RandomBlock { value, if_blocks } => {
                 // Select branch
                 let branch_val = match value {
-                    BlockValue::Random { max } => {
-                        if max == BigUint::from(0u64) {
-                            BigUint::from(0u64)
-                        } else {
-                            rng.generate(BigUint::from(1u64)..=max)
+                    BlockValue::Random { ref max } if *max == BigUint::from(0u64) => {
+                        BigUint::from(0u64)
+                    }
+                    BlockValue::Random { ref max } => {
+                        let expected: RangeInclusive<BigUint> = BigUint::from(1u64)..=max.clone();
+                        let v = rng.generate(expected.clone());
+                        if !expected.contains(&v) {
+                            warnings.push(
+                                AstParseWarning::RandomGeneratedValueOutOfRange {
+                                    expected: expected.clone(),
+                                    actual: v.clone(),
+                                }
+                                .into_wrapper_manual(0, 0),
```

Comment:
Hmm. To provide source position information of starting of a random block for warnings, it's better to wrap `Unit`'s `RandomBlock` and `SwitchBlock` struct variant with `SourcePosMixin`. Then its source position can be extracted here.